iml-agent/src/daemon_plugins/devices/devices.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    agent_error::ImlAgentError,
+    daemon_plugins::{devices::flat_devices::process_tree, DaemonPlugin, Output},
+};
+use futures::{
+    channel::oneshot, future, lock::Mutex, Future, FutureExt, Stream, StreamExt, TryFutureExt,
+    TryStreamExt,
+};
+use std::{collections::BTreeMap, pin::Pin, sync::Arc};
+use stream_cancel::{StreamExt as _, Trigger, Tripwire};
+use tokio::{
+    codec::{FramedRead, LinesCodec},
+    io::AsyncWriteExt,
+    net::UnixStream,
+};
+
+fn device_stream() -> impl Stream<Item = Result<String, ImlAgentError>> {
+    UnixStream::connect("/var/run/device-scanner.sock")
+        .err_into()
+        .and_then(|mut conn| {
+            async {
+                conn.write_all(b"\"Stream\"\n")
+                    .err_into::<ImlAgentError>()
+                    .await?;
+
+                Ok(conn)
+            }
+        })
+        .map_ok(|c| FramedRead::new(c, LinesCodec::new()).err_into())
```

Comment:
There's one device change per line? Like, one addition / removal / update.

Reply:
Currently the entire tree is emitted on a change.

`update_session` will only send the tree to the manager if it's not equal to the previous tree.